etcd/src/main/scala/io/buoyant/etcd/Etcd.scala

```diff
@@ -0,0 +1,243 @@
+  s"""$method $uri [$ps] $status"""
+})
+
+object Etcd {
+
+  private[etcd] val keysPrefixPath = Path.Utf8("v2", "keys")
+  private[etcd] val versionPath = Path.Utf8("version")
+
+  private[etcd] def mkReq(
+    path: Path,
+    method: Method = Method.Get,
+    params: Seq[(String, String)] = Nil
+  ): Request = {
+    method match {
+      case Method.Post | Method.Put =>
+        val req = Request(method, path.show)
+        req.contentType = MediaType.WwwForm
+        EtcdFormParams.set(req, params)
+        req
+
+      case _ =>
+        val req = Request(path.show, params: _*)
+        req.method = method
+        req
+    }
+  }
+
+  private[etcd] val mapper = new ObjectMapper with ScalaObjectMapper
+  mapper.registerModule(DefaultScalaModule)
+  mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+  def readJson[T: Manifest](buf: Buf): Try[T] = {
+    val Buf.ByteArray.Owned(bytes, begin, end) = Buf.ByteArray.coerce(buf)
+    Try(mapper.readValue[T](bytes, begin, end - begin))
+  }
+
+  private[etcd] def toNodeOp(
+    req: Request,
+    rsp: Response,
+    key: Path,
+    params: Seq[(String, String)] = Seq.empty
+  ): Future[NodeOp] = {
+    val state = getEtcdState(rsp)
+    (req.method, rsp.status) match {
+
+      case (Method.Get | Method.Head | Method.Delete, Status.Ok) =>
+        val op = readJson[NodeOpRsp](rsp.content).flatMap(_.toNodeOp(state))
+        Future.const(op)
+
+      case (Method.Put | Method.Post, Status.Created) =>
+        val op = readJson[NodeOpRsp](rsp.content).flatMap(_.toNodeOp(state))
+        Future.const(op)
+
+      case (method, status) =>
+        val e = readJson[ApiError](rsp.content) match {
+          case Return(apiError) => apiError
+          case Throw(e) => UnexpectedResponse(method, req.uri, params, status, state)
+        }
+        Future.exception(e)
+    }
+  }
+
+  private[etcd] object Headers {
+    val ClusterId = "x-etcd-cluster-id"
+    val EtcdIndex = "x-etcd-index"
+  }
+
+  private[etcd] def getEtcdState(msg: Message): EtcdState = {
+    val index = msg.headerMap.get(Headers.EtcdIndex).
+      flatMap { i => Try(i.toLong).toOption }.
+      getOrElse(0L)
+
+    val id = msg.headerMap.getOrElse(Headers.ClusterId, "")
+
+    EtcdState(index, id)
+  }
+}
+
+/**
+ * An etcd client.
+ *
+ * The client service is responsible for setting the Host: header, etc...
+ */
+class Etcd(client: Service[Request, Response]) extends Closable {
+
+  import Etcd._
+
+  /** Fetch etcd server version info */
+  def version(): Future[Version] = {
+    val req = mkReq(versionPath)
+    req.headerMap("accept") = MediaType.Json
+
+    client(req) flatMap { rsp =>
+      rsp.status match {
+        case Status.Ok =>
+          Future.const(readJson[Version](rsp.content))
+
+        case status =>
+          Future.exception(UnexpectedResponse(req.method, req.uri, Nil, status, EtcdState(0)))
+      }
+    }
+  }
+
+  def key(k: Path): Key = new Key(k, client)
```

Comment:
This is similar to how we designed the util-zk api at twitter.  It allows some nice separation of concerns (key selection vs actions).

For instance, if we want to "chroot" a module (i.e. lock into a prefix), we don't have to rely on the module to manage this itself (and vigilantly apply path prefixes).  We can just do:
```scala
val root = etcd.key("/some/root")
```

and then if we path the module `root`, all of its operations will occur under /some/root.

This could all use some better documentation/explanation, but I like this pattern.